src/rules/flake8_commas/rules.rs

```diff
@@ -0,0 +1,277 @@
+use itertools::Itertools;
+use rustpython_parser::lexer::{LexResult, Spanned};
+use rustpython_parser::token::Tok;
+
+use crate::ast::types::Range;
+use crate::fix::Fix;
+use crate::registry::Diagnostic;
+use crate::source_code::Locator;
+use crate::violations;
+
+/// Simplified token type.
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum TokenType {
+    Irrelevant,
+    NonLogicalNewline,
+    Newline,
+    Comma,
+    OpeningBracket,
+    OpeningSquareBracket,
+    OpeningCurlyBracket,
+    ClosingBracket,
+    For,
+    Named,
+    Def,
+    Lambda,
+    Colon,
+}
+
+/// Simplified token specialized for the task.
+#[derive(Copy, Clone)]
+struct Token<'tok> {
+    ty: TokenType,
+    // Underlying token.
+    spanned: Option<&'tok Spanned>,
+}
+
+impl<'tok> Token<'tok> {
+    fn irrelevant() -> Token<'static> {
+        Token {
+            ty: TokenType::Irrelevant,
+            spanned: None,
+        }
+    }
+
+    fn from_spanned(spanned: &'tok Spanned) -> Token<'tok> {
+        let ty = match &spanned.1 {
+            Tok::NonLogicalNewline => TokenType::NonLogicalNewline,
+            Tok::Newline => TokenType::Newline,
+            Tok::For => TokenType::For,
+            Tok::Def => TokenType::Def,
+            Tok::Lambda => TokenType::Lambda,
+            // Import treated like a function.
+            Tok::Import => TokenType::Named,
+            Tok::Name { .. } => TokenType::Named,
+            Tok::Comma => TokenType::Comma,
+            Tok::Lpar => TokenType::OpeningBracket,
+            Tok::Lsqb => TokenType::OpeningSquareBracket,
+            Tok::Lbrace => TokenType::OpeningCurlyBracket,
+            Tok::Rpar | Tok::Rsqb | Tok::Rbrace => TokenType::ClosingBracket,
+            Tok::Colon => TokenType::Colon,
+            _ => TokenType::Irrelevant,
+        };
+        Self {
+            spanned: Some(spanned),
+            ty,
+        }
+    }
+}
+
+/// Comma context type - types of comma-delimited Python constructs.
+#[derive(Copy, Clone, PartialEq, Eq)]
+enum ContextType {
+    No,
+    /// Function definition parameter list, e.g. `def foo(a,b,c)`.
+    FunctionParameters,
+    /// Call argument-like item list, e.g. `f(1,2,3)`, `foo()(1,2,3)`.
+    CallArguments,
+    /// Tuple-like item list, e.g. `(1,2,3)`.
+    Tuple,
+    /// Subscript item list, e.g. `x[1,2,3]`, `foo()[1,2,3]`.
+    Subscript,
+    /// List-like item list, e.g. `[1,2,3]`.
+    List,
+    /// Dict-/set-like item list, e.g. `{1,2,3}`.
+    Dict,
+    /// Lambda parameter list, e.g. `lambda a, b`.
+    LambdaParameters,
+}
+
+/// Comma context - described a comma-delimited "situation".
+#[derive(Copy, Clone)]
+struct Context {
+    ty: ContextType,
```

Comment:
Nit: can we call this `type_`? It's admittedly subjective, but that's the property used in several of the AST structs, so I prefer consistency.